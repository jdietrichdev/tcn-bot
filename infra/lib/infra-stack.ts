import { Construct } from 'constructs';
import { Stack, StackProps } from 'aws-cdk-lib';
import { AccessLogFormat, LogGroupLogDestination, MockIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { LogGroup } from 'aws-cdk-lib/aws-logs';

export class InfraStack extends Stack {
  private api: RestApi;
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const accessLogs = new LogGroup(this, 'access-logs', {
      logGroupName: 'BotAccessLogs',
    })

    this.api = new RestApi(this, 'discord-bot-api', {
      restApiName: 'DiscordBotApi',
      deployOptions: {
        stageName: 'prod',
        accessLogDestination: new LogGroupLogDestination(accessLogs),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields()
      }
    });

    this.api.root.resourceForPath('v1/bot').addMethod('POST', new MockIntegration({
      integrationResponses: [
        { statusCode: '200' }
      ]
    }), {
      methodResponses: [
        { statusCode: '200' },
      ]
    });
  }
}
