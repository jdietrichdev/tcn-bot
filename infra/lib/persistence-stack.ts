import { Construct } from "constructs";
import { Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { BlockPublicAccess, Bucket, BucketEncryption } from "aws-cdk-lib/aws-s3";

export class PersistenceStack extends Stack {
  readonly table: Table;
  readonly rosterTable: Table;
  readonly rosterBucket: Bucket;

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    this.table = new Table(this, "scheduling-table", {
      tableName: "SchedulingTable",
      partitionKey: {
        name: "pk",
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "sk",
        type: AttributeType.STRING,
      },
      timeToLiveAttribute: "expiration",
    });

    this.rosterTable = new Table(this, 'roster-table', {
      tableName: "RosterTable",
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: {
        name: "pk",
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "sk",
        type: AttributeType.STRING,
      },
      timeToLiveAttribute: "expires"
    });

    this.rosterBucket = new Bucket(this, 'roster-bucket', {
      bucketName: 'bot-roster-bucket',
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
    });
  }
}
