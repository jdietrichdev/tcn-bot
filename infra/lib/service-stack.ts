import { Construct } from "constructs";
import { Duration, RemovalPolicy, Stack, StackProps } from "aws-cdk-lib";
import { Code, Function as Lambda, Runtime } from "aws-cdk-lib/aws-lambda";
import {
  AccessLogFormat,
  LambdaRestApi,
  LogGroupLogDestination,
} from "aws-cdk-lib/aws-apigateway";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import {
  EventBus,
  Rule,
  RuleTargetInput,
  Schedule,
} from "aws-cdk-lib/aws-events";
import {
  CloudWatchLogGroup,
  LambdaFunction,
} from "aws-cdk-lib/aws-events-targets";
import { Table } from "aws-cdk-lib/aws-dynamodb";
import {
  BlockPublicAccess,
  Bucket,
  BucketEncryption,
  EventType,
} from "aws-cdk-lib/aws-s3";
import { LambdaDestination } from "aws-cdk-lib/aws-s3-notifications";

interface ServiceStackProps extends StackProps {
  table: Table;
  botTable: Table;
}

export class ServiceStack extends Stack {
  private eventBus: EventBus;
  private proxy: Lambda;
  private handler: Lambda;
  private scheduled: Lambda;

  constructor(scope: Construct, id: string, props: ServiceStackProps) {
    super(scope, id, props);

    const rosterBucket = new Bucket(this, "roster-bucket", {
      bucketName: "bot-roster-bucket",
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
    });

    this.handler = new Lambda(this, "bot-handler", {
      functionName: "bot-handler",
      runtime: Runtime.NODEJS_22_X,
      handler: "index.handler",
      code: Code.fromAsset("../bot/dist"),
      logRetention: RetentionDays.ONE_MONTH,
      environment: {
        REGION: props.env!.region!,
        CLASH_API_TOKEN: process.env.CLASH_API_TOKEN!,
        BOT_TOKEN: process.env.BOT_TOKEN!,
      },
      timeout: Duration.minutes(5),
      retryAttempts: 0,
    });
    props.table.grantReadWriteData(this.handler);

    this.scheduled = new Lambda(this, "bot-scheduled", {
      functionName: "bot-scheduled",
      runtime: Runtime.NODEJS_22_X,
      handler: "index.scheduled",
      code: Code.fromAsset("../bot/dist"),
      logRetention: RetentionDays.ONE_MONTH,
      environment: {
        REGION: props.env!.region!,
        CLASH_API_TOKEN: process.env.CLASH_API_TOKEN!,
        BOT_TOKEN: process.env.BOT_TOKEN!,
      },
      timeout: Duration.minutes(5),
      retryAttempts: 0,
    });

    this.eventBus = new EventBus(this, "bot-events", {
      eventBusName: "tcn-bot-events",
    });

    const eventLog = new LogGroup(this, "event-audit-log", {
      logGroupName: "EventAuditLogs",
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new Rule(this, "audit-log-rule", {
      ruleName: "AuditLogRule",
      description: "Audit log for events",
      eventBus: this.eventBus,
      eventPattern: {
        source: ["tcn-bot-event"],
      },
      targets: [new CloudWatchLogGroup(eventLog)],
    });

    new Rule(this, "bot-event-handler", {
      ruleName: "BotEventHandler",
      description: "Handler for incoming bot events",
      eventBus: this.eventBus,
      eventPattern: {
        source: ["tcn-bot-event"],
        detailType: ["Bot Event Received"],
      },
      targets: [new LambdaFunction(this.handler)],
    });

    new Rule(this, "bot-scheduled-recruiter-score", {
      schedule: Schedule.cron({
        minute: "0",
        hour: "12",
        weekDay: "SUN",
      }),
      targets: [
        new LambdaFunction(this.scheduled, {
          event: RuleTargetInput.fromObject({
            source: "tcn-bot-scheduled",
            "detail-type": "Generate Recruiter Score",
            detail: {
              guildId: "1111490767991615518",
            },
          }),
        }),
      ],
    });

    this.proxy = new Lambda(this, "bot-proxy", {
      functionName: "bot-proxy",
      runtime: Runtime.NODEJS_22_X,
      handler: "index.proxy",
      code: Code.fromAsset("../bot/dist"),
      logRetention: RetentionDays.ONE_MONTH,
      environment: {
        REGION: props.env!.region!,
        DISCORD_PUBLIC_KEY: process.env.DISCORD_PUBLIC_KEY!,
      },
    });
    this.eventBus.grantPutEventsTo(this.proxy);
    rosterBucket.grantRead(this.proxy);

    const accessLogs = new LogGroup(this, "access-logs", {
      logGroupName: "BotAccessLogs",
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new LambdaRestApi(this, "discord-bot-api", {
      restApiName: "DiscordBotApi",
      handler: this.proxy,
      deployOptions: {
        stageName: "prod",
        accessLogDestination: new LogGroupLogDestination(accessLogs),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
      },
    });

    const rosterProcessor = new Lambda(this, "roster-processor", {
      functionName: "roster-processor",
      runtime: Runtime.NODEJS_22_X,
      handler: "index.processor",
      code: Code.fromAsset("../bot/dist"),
      logRetention: RetentionDays.ONE_MONTH,
      environment: {
        REGION: props.env!.region!,
        CLASH_API_TOKEN: process.env.CLASH_API_TOKEN!,
        BOT_TOKEN: process.env.BOT_TOKEN!,
      },
      timeout: Duration.minutes(3),
    });

    rosterBucket.addEventNotification(
      EventType.OBJECT_CREATED,
      new LambdaDestination(rosterProcessor)
    );

    rosterBucket.grantRead(rosterProcessor);
    props.botTable.grantReadWriteData(rosterProcessor);
  }
}
