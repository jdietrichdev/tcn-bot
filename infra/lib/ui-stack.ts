import { Stack, StackProps } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { Nextjs } from "cdk-nextjs-standalone";

interface UIStackProps extends StackProps {
  transcriptBucket: Bucket;
}

export class UIStack extends Stack {
  readonly nextjs: Nextjs;

  constructor(scope: Construct, id: string, props: UIStackProps) {
    super(scope, id, props);

    this.nextjs = new Nextjs(this, "bot-ui", {
      nextjsPath: "../ui/",
    });

    props.transcriptBucket.grantRead(this.nextjs.serverFunction.lambdaFunction);
  }
}
