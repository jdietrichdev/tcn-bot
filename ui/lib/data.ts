import { s3Client } from "@/app/clients/s3Client";
import { PutObjectCommand } from "@aws-sdk/client-s3";

export const storeRoster = async (roster: File, server: string) => {
  try {
    const buffer = Buffer.from(await roster.arrayBuffer());
    await s3Client.send(
      new PutObjectCommand({
        Bucket: "bot-roster-bucket",
        Key: `${server}/${roster.name}`,
        Body: buffer,
        ContentType: "text/csv",
      })
    );
  } catch (err) {
    throw new Error(`Failed to store roster to bucket: ${err}`);
  }
};
