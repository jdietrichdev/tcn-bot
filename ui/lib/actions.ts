import { storeRoster } from "./data";

export type State = {
  errors?: {
    roster?: string[];
    server?: string[];
  };
  message?: string | null;
};

export async function uploadRoster(
  prevState: State,
  formData: FormData
): Promise<State> {
  const roster = formData.get("roster") as File;
  const server = formData.get("server") as string;

  await storeRoster(roster, server);

  return {
    errors: {},
    message: "Successfully uploaded roster",
  };
}
