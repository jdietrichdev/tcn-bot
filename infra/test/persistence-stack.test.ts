import { App } from "aws-cdk-lib";
import { PersistenceStack } from "../lib/persistence-stack";
import { Template } from "aws-cdk-lib/assertions";

let app: App;
let stack: PersistenceStack;
let template: string;

test("Infrastructure created", () => {
  app = new App();
  stack = new PersistenceStack(app, "stack", {
    env: { account: "12345", region: "us-east-1" },
  });
  template = JSON.stringify(Template.fromStack(stack).toJSON(), null, 2);

  expect(template).toMatchSnapshot();
});
