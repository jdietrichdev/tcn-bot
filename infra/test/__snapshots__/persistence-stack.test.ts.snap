// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infrastructure created 1`] = `
"{
  "Resources": {
    "schedulingtable85BE2571": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S"
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S"
          }
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH"
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE"
          }
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5
        },
        "TableName": "SchedulingTable",
        "TimeToLiveSpecification": {
          "AttributeName": "expiration",
          "Enabled": true
        }
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "rostertableD9ABA61B": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S"
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S"
          }
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH"
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE"
          }
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5
        },
        "SSESpecification": {
          "SSEEnabled": true
        },
        "TableName": "RosterTable",
        "TimeToLiveSpecification": {
          "AttributeName": "expires",
          "Enabled": true
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "bottableD8ACD17D": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S"
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S"
          }
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH"
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE"
          }
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5
        },
        "SSESpecification": {
          "SSEEnabled": true
        },
        "TableName": "BotTable",
        "TimeToLiveSpecification": {
          "AttributeName": "expires",
          "Enabled": true
        }
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    }
  },
  "Outputs": {
    "ExportsOutputFnGetAttrostertableD9ABA61BArnB204E024": {
      "Value": {
        "Fn::GetAtt": [
          "rostertableD9ABA61B",
          "Arn"
        ]
      },
      "Export": {
        "Name": "stack:ExportsOutputFnGetAttrostertableD9ABA61BArnB204E024"
      }
    }
  },
  "Parameters": {
    "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
    }
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5"
                  ],
                  {
                    "Ref": "BootstrapVersion"
                  }
                ]
              }
            ]
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
        }
      ]
    }
  }
}"
`;
